File: src/middleware/apiKey.ts
import { MiddlewareHandler } from "hono";
import { bearerAuth } from "hono/bearer-auth";

export const apiKeyMiddleware: MiddlewareHandler = async (c, next) => {
  const expectedApiKey = process.env.API_KEY;
  const allowQueryApiKeyInDev =
    process.env.ALLOW_QUERY_API_KEY_IN_DEV === "true";

  if (!expectedApiKey) {
    return c.json({ error: "Server misconfigured: missing API_KEY" }, 500);
  }

  const authorizationHeader = c.req.header("authorization");
  if (authorizationHeader) {
    const bearerMiddleware = bearerAuth({
      verifyToken: async (token) => token === expectedApiKey,
    });
    return bearerMiddleware(c, next);
  }

  const headerApiKey = c.req.header("x-api-key");
  const queryApiKey = allowQueryApiKeyInDev ? c.req.query("api_key") : undefined;
  const providedApiKey = headerApiKey ?? queryApiKey;

  if (!providedApiKey || providedApiKey !== expectedApiKey) {
    return c.json({ error: "Unauthorized" }, 401);
  }

  await next();
};

File: middleware/apiKey.go
package middleware

import (
	"strings"

	"github.com/gofiber/fiber/v2"
	"pixivhono/config"
)

func APIKeyMiddleware() fiber.Handler {
	return func(c *fiber.Ctx) error {
		expectedApiKey := config.GlobalConfig.APIKey
		if expectedApiKey == "" {
			return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
				"error": "Server misconfigured: missing API_KEY",
			})
		}

		authHeader := c.Get("Authorization")
		if authHeader != "" {
			// Extract token from "Bearer <token>"
			const prefix = "Bearer "
			if len(authHeader) > len(prefix) && strings.EqualFold(authHeader[:len(prefix)], prefix) {
				token := authHeader[len(prefix):]
				if token == expectedApiKey {
					return c.Next()
				}
			}
			// Hono bearerAuth middleware returns plain text "Unauthorized" on failed verify
			return c.Status(fiber.StatusUnauthorized).SendString("Unauthorized")
		}

		headerApiKey := c.Get("x-api-key")
		var queryApiKey string
		if config.GlobalConfig.AllowQueryAPIKeyInDev {
			queryApiKey = c.Query("api_key")
		}

		providedApiKey := headerApiKey
		if providedApiKey == "" {
			providedApiKey = queryApiKey
		}

		if providedApiKey == "" || providedApiKey != expectedApiKey {
			return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
				"error": "Unauthorized",
			})
		}

		return c.Next()
	}
}
